### PR TITLE
Add API for updating bootnodes and getting default ones from chainspecs

### LINKED ISSUE #981: Add bootnodes functionality API and save in extension localstorage

### PR DESCRIPTION
**Pull request by wirednkod**
- Add API for updating bootnodes and getting default ones from chainspecs
- Create Settings page with Bootnodes editing content

Fixes #981 #982

Replaces #1128

### CODE REVIEW

File: projects/extension/package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
   "dependencies": {
     "@substrate/connect-extension-protocol": "^1.0.1",
     "@substrate/smoldot-light": "0.7.2",
+    "multiaddr": "^10.0.1",
```
Comment (tomaka) [inline projects/extension/package.json:35]: I'm vetoing this dependency. The set of multiaddresses that we support is way smaller than the set of multiaddresses that this library is capable of parsing.

Please just check whether the multiaddress matches one of these two regexes:

```js
/^\/(ip4|ip6|dns4|dns6|dns)\/(.*?)\/tcp\/(.*?)\/(ws|wss|tls\/ws)\/p2p\/(.*?)$/
/^\/(ip4|ip6)\/(.*?)\/udp\/(.*?)\/webrtc\/certhash\/(.*?)\/p2p\/(.*?)$/
```

Then verify that the thing after `/p2p` is valid base58, and that the thing after `/certhash` (in the second regex) is valid base64.

File: projects/extension/src/components/Bootnodes.tsx [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    string[]
+  >([])
+
+  const [addMessage, setAddMessage] = useState<any>(undefined)
+  const [bootnodeMsgClass, setBootnodeMsgClass] = useState<string>()
+
+  useEffect(() => {
+    // Load default Bootnodes
+    const defChains = environment.getDefaultBootnodes(selectedChain)
+    setDefaultWellKnownChainBn(defChains)
```
Comment (tomaka) [inline projects/extension/src/components/Bootnodes.tsx:52]:
```suggestion
    const defChains = environment.getDefaultBootnodes(selectedChain)
    console.assert(defChains, "Invalid chain name: " + selectedChain)
    setDefaultWellKnownChainBn(defChains || [])
```

File: projects/extension/src/components/Bootnodes.tsx [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  add ? res.push(bootnode) : res.splice(res.indexOf(bootnode), 1)
+  environment.set({ type: "bootnodes", chainName }, res)
+}
+
+export const Bootnodes = () => {
+  const [selectedChain, setSelectedChain] = useState<string>("polkadot")
+  const [defaultBn, setDefaultBn] = useState<BootnodesType[]>([])
+  const [customBn, setCustomBn] = useState<BootnodesType[]>([])
+  const [customBnInput, setCustomBnInput] = useState<string>("")
+  const [defaultWellKnownChainBn, setDefaultWellKnownChainBn] = useState<
```
Comment (tomaka) [inline projects/extension/src/components/Bootnodes.tsx:42]:
```suggestion
  const [selectedChainDefaultBn, setSelectedChainDefaultBn] = useState<
```

You use the terminology "selectedChain", so let's be consistent

File: projects/extension/src/components/Bootnodes.tsx [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+      setCustomBn(tmpCust)
+    })
+  }, [selectedChain, defaultWellKnownChainBn])
+
+  const checkMultiAddr = (addr: string) => {
+    let regex58 =
+      /\/(ip4|ip6|dns4|dns6|dns)\/([a-zA-z0-9.-]{3,})\/tcp\/[0-9]{0,5}\/(ws|wss|tls|ws)\/p2p\/[a-zA-Z1-9^Il0O]{52}/i
+
+    const base64 =
+      /\/(ip4|ip6)\/([a-zA-z0-9.-]{3,})\/udp\/(.?)\/webrtc\/certhash\/(.*?)\/p2p\/[-A-Za-z0-9+=]{1,50}|=[^=]|={3,}/i
```
Comment (tomaka) [inline projects/extension/src/components/Bootnodes.tsx:96]: Why is the thing after `/p2p/` different than for `regex58`? It's exactly the same format

File: projects/extension/src/components/Bootnodes.tsx [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+        })
+      }
+      setDefaultBn(tmpDef)
+      setCustomBn(tmpCust)
+    })
+  }, [selectedChain, defaultWellKnownChainBn])
+
+  const checkMultiAddr = (addr: string) => {
+    let regex58 =
+      /\/(ip4|ip6|dns4|dns6|dns)\/([a-zA-z0-9.-]{3,})\/tcp\/[0-9]{0,5}\/(ws|wss|tls|ws)\/p2p\/[a-zA-Z1-9^Il0O]{52}/i
```
Comment (tomaka) [inline projects/extension/src/components/Bootnodes.tsx:93]:
```suggestion
      /\/(ip4|ip6|dns4|dns6|dns)\/([a-zA-z0-9.-]{3,})\/tcp\/[0-9]{0,5}\/(ws|wss|tls\/ws)\/p2p\/[a-zA-Z1-9^Il0O]{52}/i
```

File: projects/extension/src/components/Bootnodes.tsx [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+    const base64 =
+      /\/(ip4|ip6)\/([a-zA-z0-9.-]{3,})\/udp\/(.?)\/webrtc\/certhash\/(.*?)\/p2p\/[-A-Za-z0-9+=]{1,50}|=[^=]|={3,}/i
+
+    if (!regex58.test(addr) || !regex58.test(addr))
+      throw new Error(
+        "Multiaddress provided is not correct (not base58 or base64 format).",
```
Comment (tomaka) [inline projects/extension/src/components/Bootnodes.tsx:100]:
```suggestion
        "Multiaddress provided is not correct.",
```

Your regexes don't check for the base58 or base64 formats

File: projects/extension/src/components/Bootnodes.tsx [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+import "./Bootnodes.css"
+import { Title, Switch } from "."
+interface BootnodesType {
+  checked: boolean
+  bootnode: string
+}
+
+interface Boot {
+  [key: string]: string[]
+}
```
Comment (tomaka) [inline projects/extension/src/components/Bootnodes.tsx:14]: This seems to be unused?

File: projects/extension/src/components/Bootnodes.tsx (hunk 4 above)
Comment (wirednkod) [reply projects/extension/src/components/Bootnodes.tsx:96]: I was under the impression that the bootnode's structure of WebRTC is different. I think its better to alter those to match [the ones in smoldot](https://github.com/paritytech/smoldot/blob/main/bin/wasm-node/javascript/src/index-browser.ts#L88-L90)

File: projects/extension/src/components/Bootnodes.tsx [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        })
+      }
+      setDefaultBn(tmpDef)
+      setCustomBn(tmpCust)
+    })
+  }, [selectedChain, selectedChainDefaultBn])
+
+  const checkMultiAddr = (addr: string) => {
+    const ws =
+      /\/(ip4|ip6|dns4|dns6|dns)\/([a-zA-z0-9.-]{3,})\/tcp\/[0-9]{0,5}\/(ws|wss|tls\/ws)\/p2p\/[a-zA-Z1-9^Il0O]{52}/i
```
Comment (tomaka) [inline projects/extension/src/components/Bootnodes.tsx:90]:
```suggestion
      /\/(ip4|ip6|dns4|dns6|dns)\/(.?)\/tcp\/[0-9]{0,5}\/(ws|wss|tls\/ws)\/p2p\/[a-zA-Z1-9^Il0O]{52}/i
```

Domain names are surprisingly complicated

Aside from this, I'm not super sure that the peer ID is always 52 characters long, but I'm too lazy to check this

File: projects/extension/src/components/Bootnodes.tsx [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+    const webrtc =
+      /\/(ip4|ip6)\/([a-zA-z0-9.-]{3,})\/udp\/(.?)\/webrtc\/certhash\/(.*?)\/p2p\/[a-zA-Z1-9^Il0O]{52}/i
```
Comment (tomaka) [inline projects/extension/src/components/Bootnodes.tsx:93]:
```suggestion
      /\/(ip4|ip6)\/(.?)\/udp\/(.?)\/webrtc\/certhash\/(.*?)\/p2p\/[a-zA-Z1-9^Il0O]{52}/i
```

**Review by tomaka — APPROVED** [review]

**Pull request merged**